### PR TITLE
Enable Widevine on all Linux archs

### PR DESCRIPTION
**Pull request by mherrmann**
Resolves https://github.com/brave/brave-browser/issues/33081.

After this PR, Arm64 Linux users can set up Widevine as follows:

1. Obtain a copy of Widevine. This is a folder with the files `manifest.json` and `_platform_specific/linux_arm64/libwidevinecdm.so`. Some Linux distributions such as Raspberry Pi OS ship with a copy of Widevine somewhere in `/opt`.
2. In Brave's User Data Directory, create the file `WidevineCdm/latest-component-updated-widevine-cdm` with the following contents: `{"Path":"/path/to/the/folder/from/step/1"}`. The default User Data Directory for the Release version of Brave is `~/.config/BraveSoftware/Brave-Browser`.
3. Enable Widevine in `brave://settings`.
4. Restart Brave.

The instructions can also be used on architectures other than Arm64. In those cases, the path that mentions `linux_arm64` above will differ.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] **Update brave.com with the instructions above.**
- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

At a minimum, it should be tested that Widevine can now be used on Arm64 Linux. The expected behavior is that Widevine should "just work" when users follow the steps at the top of this PR description. The only real feasible platform I have found for testing the functionality is on a Raspberry Pi 4B.

If you don't have a physical Raspberry Pi, then it is also possible to emulate it with QEMU. [This article](https://interrupt.memfault.com/blog/emulating-raspberry-pi-in-qemu) gives good instructions. Note the `-nographic` comment at the bottom. However, setting this up is very tedious and the emulation is extremely slow. I would advise to buy a physical Raspberry Pi for testing instead.

The changes may also affect Widevine on other platforms. If there is time, then it would be good to test that Widevine can still be installed and used to play DRM-protected content on other platforms.

### CODE REVIEW

**Comment by theofficialgman** [comment]
> @theofficialgman do you have advice (for testing) what the easiest way of obtaining Widevine on some Arm64 Linux distribution is?

@mherrmann the easiest distribution to test on would likely be Raspberry PiOS Bookworm ARM64 (on a pi3/pi4/pi5) as they already have the widevine GLIBC patches/hacks implemented by default as well as ship the patched ARM64 chromeOS widevine binaries in `/opt/WidevineCdm` from their `libwidevinecdm` package

There are some other ARM64 distros as well that do this but on more exotic hardware (eg: Nintendo Switch)

**Comment by mherrmann** [comment]
@theofficialgman OK, thank you!

**Comment by simonhong** [comment]
It looks like this PR assumes that linux on arm64 pre-installs widevine binary duing the installation not by component update with user consent.
AFAIK, we don't want bundling proprietary binary. maybe we need security review about this approach.
Or already discussed this way? cc @diracdeltas

**Comment by diracdeltas** [comment]
> It looks like this PR assumes that linux on arm64 pre-installs widevine binary duing the installation not by component update with user consent. AFAIK, we don't want bundling proprietary binary. maybe we need security review about this approach. Or already discussed this way? cc @diracdeltas

i haven't discussed this with anyone. however @mherrmann said

> users can place the Widevine files into WidevineCdm/ in Brave's installation directory to get Widevine to work on architectures where it isn't publicly available.

if users have to manually download widevine files into a directory to get it to work, i'm not sure the consent popup is needed. cc @bbondy

**Comment by mherrmann** [comment]
Thank you for taking a look at my PR @simonhong. I wouldn't say the PR _assumes_ that the Widevine binaries are pre-installed. If they are not pre-installed then Brave works just like it did before. I would say that users _may_ pre-install the Widevine binaries. In this case, you are right that I have assumed user consent, and opted not to show another popup. (It was also easier to implement in this way.) And you're also right that it's good to confirm whether others think that it is OK to assume that. :+1:

File: browser/brave_tab_helpers.cc
Comment (simonhong) [inline browser/brave_tab_helpers.cc:163]: Instead of using `#ifdef`, how about using something like `BraveDrmTabHelper::MaybeCreateForWebContents()`?
and that static method could be no op if it's arm64 linux.

File: browser/brave_content_browser_client.cc
Comment (simonhong) [inline browser/brave_content_browser_client.cc:582]: ditto - how about using something like `bool NeedBraveDrmTabHelper()` util method?
Then, we could remove most of ifdefs from source and gn files.

File: browser/widevine/widevine_utils.cc
Comment (simonhong) [inline browser/widevine/widevine_utils.cc:158]: Just curious `bundled_dir.IsParent(it.path)` is necessary here. checking `it.key_system` isn't sufficient?

File: browser/widevine/widevine_utils.cc
Comment (simonhong) [inline browser/widevine/widevine_utils.cc:191]: How about deleting MigrateWidevinePrefs() instead of adding more code to it because this migration code was added almost 3y ago. I think we could safely delete all methods that having Migrate or Obsolete in their name.
Maybe this code could be added in `RegisterWidevineLocalstatePrefs()`.

File: chromium_src/chrome/common/media/cdm_registration.cc
Comment (simonhong) [inline chromium_src/chrome/common/media/cdm_registration.cc:12]: We disabled `bundle_widevine_cdm` by default. `WidevineBuildFlag.FlagTest` test case checks it.
I think we don't need this file?

File: browser/widevine/widevine_utils.h
Comment (simonhong) [inline browser/widevine/widevine_utils.h:39]: this method name made me confusing :) Maybe using `HasCopiedWideve()` more intuitive. WDYT?

File: browser/widevine/widevine_utils.cc
Comment (mherrmann) [reply browser/widevine/widevine_utils.cc:158]: In general, `cdm_registry->GetRegisteredCdms()` also contains component-updated Widevine. The `IsParent` check is needed to distinguish a component-updated Widevine from a bundled one.

On the other hand, you are right that at the moment, on non-x64 Linux where there is currently no component-updated Widevine, the `IsParent` is not necessary. But maybe in the future there will be component-updated Widevine on Arm64 Linux. Then, the `IsParent` check will matter.

File: browser/widevine/widevine_utils.h
Comment (mherrmann) [reply browser/widevine/widevine_utils.h:39]: I see; What do you think of `IsWidevinePreinstalled`? I'm suggesting this alternative because `Copied` sounds very general to me.

Comment (simonhong) [reply browser/widevine/widevine_utils.h:39]: `IsWidevinePreinstalled()` looks good!

File: chromium_src/chrome/common/media/cdm_registration.cc
Comment (mherrmann) [reply chromium_src/chrome/common/media/cdm_registration.cc:12]: I understand about `bundle_widevine_cdm`. But does it really mean that this whole file `cdm_registration.cc` is not needed? I believe we still need to replace `BUILDFLAG_INTERNAL_BUNDLE_WIDEVINE_CDM` by `BUILDFLAG_INTERNAL_ENABLE_WIDEVINE_CDM_COMPONENT`. As a related note, this comes from [snap-from-source 56c42943](https://git.launchpad.net/~chromium-team/chromium-browser/+git/snap-from-source/tree/build/chromium-patches/widevine-enable-version-string.patch?id=56c42943).

File: browser/widevine/widevine_utils.h
Comment (mherrmann) [reply browser/widevine/widevine_utils.h]: Fixed in b1705af42913eaa12ef852d06dc47a008e6e73be.

File: chromium_src/chrome/common/media/cdm_registration.cc
Comment (simonhong) [reply chromium_src/chrome/common/media/cdm_registration.cc:12]: As we already use widevine cdm component on linux, I thought we could use widevine cdm component's directory for our pre-installed location instead of using additional bundle define.
According to the below code, I assume that pre-installed widevine binary would work if it's stored in component directory?
```
#if BUILDFLAG(BUNDLE_WIDEVINE_CDM)
  bundled_widevine = GetBundledWidevine();
#endif

  content::CdmInfo* updated_widevine = nullptr;
#if BUILDFLAG(ENABLE_WIDEVINE_CDM_COMPONENT)
  updated_widevine = GetComponentUpdatedWidevine();
#endif
```

Comment (simonhong) [reply chromium_src/chrome/common/media/cdm_registration.cc:12]: With that, I think we could use component updater for arm64 w/o more work in the future.

Comment (mherrmann) [reply chromium_src/chrome/common/media/cdm_registration.cc]: > According to the below code, I assume that pre-installed widevine binary would work if it's stored in component directory?

It is complicated to set up, but you are right. I will use this point to significantly simplify the PR.

**Comment by mherrmann** [comment]
(I removed the `CI/run-linux-arm64` label because CI is currently broken for Arm64 Linux. See https://github.com/brave/brave-browser/issues/34984.)

**Comment by mherrmann** [comment]
@simonhong I found a simpler implementation that lets users on non-x64 Linux install Widevine into the standard component directory. (I updated the instructions for doing so in the PR description at the top and will want to add these instructions to brave.com.) I also found a convenient single place for turning `BraveDrmTabHelper` into a no-op. I have a slight preference for the `#if` and conditional compilation approach I had before but am reasonably happy with this new implementation as well. Please let me know what you think.

**Comment by mherrmann** [comment]
@bbondy the implementation has changed so the question asked of you above has become obsolete. The latest implementation no longer assumes consent when the user installs Widevine manually.

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (bridiver) [inline patches/third_party-widevine-cdm-widevine.gni.patch:9]: can you submit an upstream CL to put `library_widevine_cdm_available` in `declare_args`?

Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch:9]: I have started the process of getting `library_widevine_cdm_available` into `declare_args` in upstream but do not know whether or when it would land. Can we merge this PR as-is? I'll follow through on getting the work into upstream (and then hopefully to update our implementation to use the GN arg). But until then, we will have already delivered value to users.

Comment (goodov) [reply patches/third_party-widevine-cdm-widevine.gni.patch:9]: > I have started the process of getting `library_widevine_cdm_available` into `declare_args` in upstream but do not know whether or when it would land. Can we merge this PR as-is? I'll follow through on getting the work into upstream (and then hopefully to update our implementation to use the GN arg). But until then, we will have already delivered value to users.

if the CL is in the works, pls consider updating this PR to have a patch that adds `declare_args() { }` around `library_widevine_cdm_available` declaration and passing `library_widevine_cdm_available=true` as a gn arg if `targetOS == linux`.

Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch:9]: The CL is in the works in the following sense: I've never contributed to Chromium. I tried to create an issue on https://bugs.chromium.org/p/chromium/issues/wizard a few hours ago. The wizard ended with a message along the lines of "thanks, it will take some time for the issue to appear". I didn't receive a confirmation email or anything similar. I suspect the issue is in a review queue to prevent spam. I also have a local branch with the necessary changes. Then I believe I will need to `git cl upload` (and probably some other commands) to start the review process.

Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: @goodov I implemented your suggestion in 8f475ce95bfc61dd4f081d90790721859aff77f0. Wdyt?

Comment (bridiver) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: @mherrmann the idea is that it should be exactly the patch that gets submitted to upstream. That way it just goes away on its own when we update chromium. There should be no additional comments, etc...

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (goodov) [inline patches/third_party-widevine-cdm-widevine.gni.patch:9]: ?

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (goodov) [inline patches/third_party-widevine-cdm-widevine.gni.patch:13]: as Brian mentioned, no comments needed, so just `declare_args() {`.

**Review by goodov — APPROVED** [review]
lgtm with few patch nits.

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: Resolved in 07e9aeebff86e2f79ba0b39aa06bdb6291f39659.

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: Resolved in 07e9aeebff86e2f79ba0b39aa06bdb6291f39659.

File: patches/third_party-widevine-cdm-widevine.gni.patch
Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: I've submitted the patch upstream at https://chromium-review.googlesource.com/c/chromium/src/+/5186136 and am hoping for their review.

Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch]: Upstream's initial response was to reject my patch:

> Widevine CDM isn't supported on Linux arm64 so I am afraid we cannot approve this cahnge.

I'm still in touch with them to try to convince them that this change makes sense.

Comment (mherrmann) [reply patches/third_party-widevine-cdm-widevine.gni.patch:9]: @bridiver I don't think I will manage to convince upstream to `put library_widevine_cdm_available` in `declare_args` so I undid my changes to that effect. Can we merge this PR as-is? If upstream do decide to accept the GN arg, then I will switch to the GN arg based implementation in a follow-up PR.

**Review by bridiver — APPROVED** [review]

**Comment by theofficialgman** [comment]
Just wanted to add additional info.
Newer distros (on GLIBC 2.36+) no longer need to patch GLIBC at all. All that is needed is to patch Widevine as detailed here https://github.com/theofficialgman/l4t-debs/commit/09c7234202a29b5803f68f456ab9503600aa4722

You can download a pre-patched version here -> https://github.com/theofficialgman/testing/releases/download/gmans-releases/WidevineCdm-2.36.tar.gz . Example distros where this works are Debian 12+ and Ubuntu 23.10+

**Comment by mherrmann** [comment]
@theofficialgman I'm not a lawyer but fear you may be breaking Widevine's license by re-distributing it. I wish everybody could distribute it. But for your protection, I recommend considering removing those binaries.

**Comment by theofficialgman** [comment]
> @theofficialgman I'm not a lawyer but fear you may be breaking Widevine's license by re-distributing it. I wish everybody could distribute it. But for your protection, I recommend considering removing those binaries.

I am also not a lawyer however....

First, there is no associated license file with the chromeOS WidevineCDM.

raspberry pi LTD currently packages `libwidevinecdm0` in their public apt repository of similar contents, maintainer David Turner.

In addition to raspberry pi LTD, packaging a redistribution of WidevineCDM has already been adopted by those that redistribute chromeOS recovery images as well as those that repackage chrome Linux (eg: chrome on flathub).

Thus, from my perspective, everyone is redistributing and likely can redistribute.

**Comment by theofficialgman** [comment]
@mherrmann any chance we could get these changes picked to stable given that we are still a month out to 1.63.x hitting stable?

**Comment by mherrmann** [comment]
@theofficialgman no, sorry. That's only possible for critical fixes. Letting changes ride the nightly -> beta -> stable train reduces the risk that they break something before we find out.

**Comment by mherrmann** [comment]
We now have official instructions for enabling Widevine on Arm64 Linux here: https://support.brave.com/hc/en-us/articles/23881756488717-How-do-I-enable-Widevine-DRM-on-Linux

**Comment by theofficialgman** [comment]
> We now have official instructions for enabling Widevine on Arm64 Linux here: https://support.brave.com/hc/en-us/articles/23881756488717-How-do-I-enable-Widevine-DRM-on-Linux

@mherrmann thanks for the official instructions. I figured them out through commit history previously.
There seems to be no way to set a global `latest-component-updated-widevine-cdm` file (eg: for all current users and users that might get made on a system in the future). Is this going to be a supported usecase in the future? I know that chromium has methods for setting defaults globally (eg: preferences, bookmarks, and extensions) for all users but it appears that brave settings and files such as `latest-component-updated-widevine-cdm` have no way to set these defaults globally.

**Comment by mherrmann** [comment]
@theofficialgman maybe it already works if you copy `WidevineCdm` into Brave's installation directory? If not, you can create a new issue at https://github.com/brave/brave-browser/issues/ to request this as a feature.

**Comment by theofficialgman** [comment]
> @theofficialgman maybe it already works if you copy `WidevineCdm` into Brave's installation directory? If not, you can create a new issue at https://github.com/brave/brave-browser/issues/ to request this as a feature.

@mherrmann actually that works. brave automatically populates the `latest-component-updated-widevine-cdm` file with `{"Path":"/opt/brave.com/brave/WidevineCdm"}` if I have widevine present already in that directory before enabling with the extention.

**Pull request merged**
